### PR TITLE
feat(repeat,yoyo

### LINKED ISSUE #492: feat(motion): add repeat, loop, and yoyo modifiers
> For AI agents (Claude Code, Copilot, Cursor): read [AGENTS.md](https://github.com/Karanjot786/TermUI/blob/main/AGENTS.md) and [packages/motion/AGENTS.md](https://github.com/Karanjot786/TermUI/blob/main/packages/motion/AGENTS.md) before writing code. This issue is a complete spec. Follow it exactly.

## Objective

Add a `repeat` wrapper around an `AnimationRunner`. It re-runs the runner a fixed number of times, loops forever, and optionally reverses direction each pass (yoyo).

## Package scope

`packages/motion`. All work is confined here. Import shared types from `@termuijs/core`.

## Files to create or modify

```
MODIFY: packages/motion/src/sequence.ts   (add repeat)
CREATE: packages/motion/src/repeat.test.ts
MODIFY: packages/motion/src/index.ts       (export repeat, RepeatOptions)
```

## API contract

```typescript
import type { AnimationRunner } from './sequence.js';

export interface RepeatOptions {
  /** Number of passes. Use Infinity for an endless loop. Default 1. */
  count?: number;
  /** Reverse direction on every other pass. Default false. */
  yoyo?: boolean;
}

export function repeat(
  runner: AnimationRunner,
  options?: RepeatOptions,
): AnimationRunner;
```

## Acceptance criteria

- [ ] `repeat(r, { count: 3 })` runs the runner 3 times then calls its own `done`
- [ ] `count: 1` (default) runs the runner exactly once
- [ ] `count: Infinity` never calls `done` until cancelled
- [ ] The returned cancel function stops the active pass and prevents further passes
- [ ] With `yoyo: true` even passes run the runner forward and odd passes run a reversed runner
- [ ] `repeat` is a pure wrapper and starts no timers itself
- [ ] `bun vitest run packages/motion` passes
- [ ] `bun run typecheck` passes

## Test expectations

```
File: packages/motion/src/repeat.test.ts
Run:  bun vitest run packages/motion
```

Named cases:
- `runs the runner count times`
- `default count runs once`
- `cancel stops further passes`
- `infinite loop does not call done`
- `yoyo alternates direction each pass`

Use a synchronous fake runner: `const fake: AnimationRunner = (done) => { calls++; done(); return () => {}; }`. Count invocations with a plain counter or `vi.fn()`. Never mutate shared state.

## Reference pattern to follow

Follow `packages/motion/src/sequence.ts`. The `sequence` function already chains runners with a `runNext(index)` recursion and a `cancelled` flag plus `cancelCurrent`. Mirror that control flow.

## Not included

- Delay between passes
- Per-pass easing changes
- Parallel repeat (compose with `parallel` instead)

## Do not touch

- Any package other than `packages/motion`
- `packages/core` (import shared types, do not modify them)
- `bun.lock`
- `.github/`

## Agent prompt (copy this into your AI agent)

```
Read AGENTS.md at the repo root and packages/motion/AGENTS.md.
Read the reference file packages/motion/src/sequence.ts and copy its structure.
Task: add repeat(runner, options) exactly to the API contract in this issue.
Plan first. Reuse the cancelled/cancelCurrent control flow from sequence(). Export repeat and RepeatOptions from index.ts.
Confine all work to packages/motion. Do not edit bun.lock. Do not add dependencies.
Tests use synchronous fake runners; no real timers.
When done run: bun vitest run packages/motion && bun run typecheck. Both must pass.
```

## Before you open the PR

- [ ] I run `bun run build && bun vitest run && bun run typecheck` and all pass
- [ ] My change is confined to `packages/motion`
- [ ] `bun.lock` has no unrelated changes
- [ ] I starred the repo

## GSSoC 2026

This is an intermediate-level issue. Comment "I would like to work on this" to get assigned. You have 7 days to open a PR after assignment.

### PR DESCRIPTION
**Pull request by Krishnavamsi-codes**
## Description

Adds a `repeat()` wrapper for `AnimationRunner` in `@termuijs/motion`.

The wrapper supports fixed repetition counts, infinite looping via `count: Infinity`, cancellation of active/future passes, and optional yoyo behavior that alternates between forward and reversed runners. Added comprehensive tests covering all required repeat scenarios.

## Related Issue

Closes #492

## Which package(s)?

* @termuijs/motion

## Type of Change

* [x] ✨ Feature (`type:feature`)
* [x] 🧪 Tests (`type:testing`)

## Checklist

* [x] ⭐ You starred the repo. The `needs-star` check blocks your merge otherwise.
* [x] Tests pass locally: `bun vitest run`
* [x] Build passes: `bun run build`
* [x] Typecheck passes: `bun run typecheck`
* [x] You read [`[CONTRIBUTING.md](https://chatgpt.com/g/g-p-6a1c81640ef4819180e1541a1faf22fa-oss-termui/c/CONTRIBUTING.md)`](./CONTRIBUTING.md).
* [x] Your PR title follows `type: short description`.
* [ ] Widget state mutators call `markDirty()` (if your change affects rendering).
* [x] No new `any` types without an inline comment explaining why.
* [x] No unrelated refactors bundled into this PR.

## GSSoC 2026 Participation

* [x] You are a GSSoC 2026 contributor.
* Your GSSoC profile: https://gssoc.girlscript.org/profile/YOUR_PROFILE_ID`](https://gssoc.girlscript.org/profile/42e0e4bf-0550-4021-a19c-29f6f92ca35d

## Screenshots / Recordings (UI changes)

N/A

### Changes

* Added `RepeatOptions` interface.
* Added `repeat(runner, options)` implementation in `packages/motion/src/sequence.ts`.
* Exported `repeat` and `RepeatOptions` from `packages/motion/src/index.ts`.
* Added `repeat.test.ts` covering:

  * runs the runner count times
  * default count runs once
  * cancel stops further passes
  * infinite loop does not call done
  * yoyo alternates direction each pass

### Implementation Notes

* Mirrors the recursive control flow used by `sequence()`.
* Uses the same `cancelled` and `cancelCurrent` pattern for cancellation.
* Starts no timers and acts as a pure wrapper around existing runners.
* For yoyo mode, a reversed runner is used when supplied; otherwise execution falls back to the original runner.

## Summary by CodeRabbit

* **New Features**
  * Added a repeat helper to run animations multiple times with configurable count and optional yoyo (alternate direction) behavior.

* **Tests**
  * Added comprehensive tests covering counts, default behavior, cancellation, infinite runs, and yoyo edge cases.

### CODE REVIEW

File: packages/motion/src/sequence.ts [hunk 1]
```diff
@@ -75,4 +75,53 @@ export function parallel(
...
+      done()
+      return () => {}
+    }
+
+    let cancelled = false
+    let cancelCurrent: () => void = () => {}
+
+    function runNext(index: number) {
+      if (cancelled || index >= count) {
+        if (!cancelled) done()
```
[the remark this answers is not in the document]
Comment (Krishnavamsi-codes) [reply packages/motion/src/sequence.ts]: Thanks for the suggestion. I kept the implementation aligned with the issue specification, which doesn't define validation or error handling for invalid `count` values. To avoid introducing additional API behavior beyond the scope of this issue, I left the current behavior unchanged.
